File: src/main/java/com/hileco/cortex/vm/layer/LayeredMap.java
package com.hileco.cortex.vm.layer;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class LayeredMap<K, V> implements MapApi<K, V, LayeredMap<K, V>> {
    private LayeredMap<K, V> parent;
    private Map<K, V> layer;
    private Set<K> deletions;

    public LayeredMap() {
        this.parent = null;
        this.layer = new HashMap<>();
        this.deletions = new HashSet<>();
    }

    private LayeredMap(LayeredMap<K, V> parent, Map<K, V> layer, Set<K> deletions) {
        this.parent = parent;
        this.layer = layer;
        this.deletions = deletions;
    }

    @Override
    public LayeredMap<K, V> copy() {
        if (!this.layer.isEmpty() || !this.deletions.isEmpty()) {
            this.parent = new LayeredMap<>(this.parent, this.layer, this.deletions);
            this.layer = new HashMap<>();
            this.deletions = new HashSet<>();
        }
        return new LayeredMap<>(this.parent, new HashMap<>(), new HashSet<>());
    }

    @Override
    public int size() {
        return this.keySet().size();
    }

    @Override
    public boolean isEmpty() {
        return this.layer.isEmpty() && (this.parent == null || this.parent.isEmpty());
    }

    @Override
    public boolean containsKey(K key) {
        return this.layer.containsKey(key) || (this.parent != null && this.parent.containsKey(key) && !this.deletions.contains(key));
    }

    @Override
    public V get(K key) {
        if (this.layer.containsKey(key)) {
            return this.layer.get(key);
        }
        if (this.parent != null && !this.deletions.contains(key)) {
            return this.parent.get(key);
        }
        return null;
    }

    @Override
    public V put(K key, V value) {
        var previous = this.get(key);
        this.deletions.remove(key);
        this.layer.put(key, value);
        return previous;
    }

    public <K2 extends K, V2 extends V> void merge(LayeredMap<K2, V2> map) {
        map.keySet().forEach(key -> {
            var value = map.get(key);
            this.put(key, value);
        });
    }

    @Override
    public V remove(K key) {
        var previous = this.get(key);
        this.layer.remove(key);
        if (this.parent != null && this.parent.containsKey(key)) {
            this.deletions.add(key);
        }
        return previous;
    }

    @Override
    public void clear() {
        this.parent = null;
        this.layer.clear();
        this.deletions.clear();
    }

    @Override
    public Set<K> keySet() {
        HashSet<K> keys = new HashSet<>();
        if (this.parent != null) {
            keys.addAll(this.parent.keySet());
        }
        keys.addAll(this.layer.keySet());
        if (this.parent != null) {
            keys.removeAll(this.deletions);
        }
        return keys;
    }

    @Override
    public String toString() {
        var stringBuilder = new StringBuilder();
        stringBuilder.append("LayeredMap{");
        stringBuilder.append("\n");
        this.keySet().forEach(key -> {
            stringBuilder.append("[");
            stringBuilder.append(key);
            stringBuilder.append("] = ");
            stringBuilder.append(this.get(key));
            stringBuilder.append(",");
            stringBuilder.append("\n");
        });
        stringBuilder.append("}");
        return stringBuilder.toString();
    }
}

File: src/main/java/com/hileco/cortex/vm/layer/LayeredMap.kt
package com.hileco.cortex.vm.layer

class LayeredMap<K, V>(parent: LayeredMap<K, V>? = null) : TreeLayered<LayeredMap<K, V>>(parent) {
    private var layer: MutableMap<K, V> = HashMap()
    private var deletions: MutableSet<K> = HashSet()

    init {
        val mergeableParent = this.parent
        if (mergeableParent != null
                && mergeableParent.layer.size < LayeredMap.MINIMUM_LAYER_SIZE
                && mergeableParent.deletions.size < LayeredMap.MINIMUM_DELETIONS_SIZE) {
            mergeParent(mergeableParent)
            this.parent = mergeableParent.parent()
        }
    }

    @Synchronized
    override fun extractParentLayer(parent: LayeredMap<K, V>?): LayeredMap<K, V> {
        val clone = LayeredMap(parent)
        clone.layer = layer.toMutableMap()
        clone.deletions = deletions.toMutableSet()
        return clone
    }

    @Synchronized
    override fun isLayerEmpty(): Boolean {
        return layer.isEmpty() && deletions.isEmpty()
    }

    @Synchronized
    override fun createSibling(parent: LayeredMap<K, V>?): LayeredMap<K, V> {
        return LayeredMap(parent)
    }

    @Synchronized
    override fun mergeParent(parent: LayeredMap<K, V>) {
        synchronized(parent) {
            parent.layer.forEach { key, value ->
                if (!deletions.contains(key)) {
                    layer.putIfAbsent(key, value)
                }
            }
            parent.deletions.forEach { deletion ->
                if (!layer.containsKey(deletion)) {
                    deletions.add(deletion)
                }
            }
        }
    }

    fun size(): Int {
        return keySet().size
    }

    @Synchronized
    operator fun get(key: K): V? {
        if (layer.containsKey(key)) {
            return layer[key]
        }
        return if (!deletions.contains(key)) {
            return parent?.get(key)
        } else null
    }

    @Synchronized
    operator fun set(key: K, value: V) {
        deletions.remove(key)
        layer[key] = value
    }

    @Synchronized
    fun remove(key: K) {
        layer.remove(key)
        val currentParent = parent
        if (currentParent?.get(key) != null) {
            deletions.add(key)
        }
    }

    @Synchronized
    fun keySet(): Set<K> {
        val keys = HashSet<K>()
        val currentParent = parent
        if (currentParent != null) {
            keys.addAll(currentParent.keySet())
        }
        keys.addAll(layer.keys)
        keys.removeAll(deletions)
        return keys
    }

    override fun toString(): String {
        val stringBuilder = StringBuilder()
        stringBuilder.append("LayeredMap{")
        stringBuilder.append("\n")
        keySet().forEach { key ->
            stringBuilder.append("[")
            stringBuilder.append(key)
            stringBuilder.append("] = ")
            stringBuilder.append(this[key])
            stringBuilder.append(",")
            stringBuilder.append("\n")
        }
        stringBuilder.append("}")
        return "$stringBuilder"
    }

    override fun equals(other: Any?): Boolean {
        return other is LayeredMap<*, *> && let {
            other as LayeredMap<K, V>
            val keySet = keySet()
            keySet == other.keySet() && keySet.all { key ->
                val thisValue = this[key]
                val otherValue = other[key]
                return (thisValue == null && otherValue == null)
                        || (thisValue != null && thisValue == otherValue)
            }
        }
    }

    override fun hashCode(): Int {
        return keySet().hashCode()
    }

    companion object {
        const val MINIMUM_LAYER_SIZE: Int = 4
        const val MINIMUM_DELETIONS_SIZE: Int = 4
    }
}
